File: src/examples/python/camera.py
from migo import *
import cv2
import math as m

handler = connect()
trace(handler, True)

MaxV = 5.0
road_color_min = np.array([0, 0, 0],np.uint8)
road_color_max = np.array([255, 90, 180],np.uint8)

try:
    while True:
        sz, buff = captureCamera(handler)
        img = cameraToCV(buff)

        hsv = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
        mask = cv2.inRange(hsv, road_color_min, road_color_max)
        lineMask = mask[150:160,:]

        cnts = cv2.findContours(lineMask.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        cnts = cnts[1]
        w = 0
        if len(cnts) > 0:
            c = cnts[0]
            M = cv2.moments(c)
            if M["m00"] != 0:
                cX = int(M["m10"] / M["m00"])
                cY = int(M["m01"] / M["m00"])


                cv2.line(img, (cX, 155 + cY), (160, 240), (0,0,255), 3)
                #cv2.circle(img, (cX, 155 + cY), 7, (255, 255, 255), -1)
                w = 0.6*m.atan2(160-cX, 80)
                print(m.degrees(w))

        

        velocity(handler, [MaxV, w])

        cv2.imshow("image", img)
        cv2.imshow("mask", mask)
        key = cv2.waitKey(1)
        if key == 27:
            break
except Exception, e:
    print(e)

trace(handler, False)
disconnect(handler)

File: src/examples/matlab/camera.m
clear; close all; clc;
disconnectRobotOne(1);

handler = connectRobotOne('127.0.0.1');
traceRobotOne(handler, true);

MaxV = 5.0;

for n = 1:1000
   rgb = captureCamera(handler);
   hsv = rgb2hsv(rgb);
   
   minval = [0 0 0]/255;
   maxval = [255 90 180]/255;

   mask = true(size(hsv,1), size(hsv,2));
   for p = 1 : 3
        mask = mask & (hsv(:,:,p) >= minval(p) & hsv(:,:,p) <= maxval(p));
   end
   
   lineMask = mask(150:160, :);
   [row, col] = find(lineMask);
   
   w = 0;
   if length(row) > 0
        mrow = sum(row)/length(row);
        mcol = sum(col)/length(col);
        x = mcol;
        y = mrow + 150;
        w = 0.6*atan2(160-x, 80);
        disp(w)
   end
   velocity(handler, [MaxV w]);
   
   figure(1)
   imshow(rgb)
   figure(2)
   imshow(mask)
   waitRobotOne(handler)
end

traceRobotOne(handler, false);
disconnectRobotOne(handler);